File: src/Builtins.hs
module Builtins where

import Parse.Syntax



builtins :: [(Name, Type)]
builtins = [
    ("+", TypeFunction [TypeInt, TypeInt] TypeInt),
    ("-", TypeFunction [TypeInt, TypeInt] TypeInt),
    ("<", TypeFunction [TypeInt, TypeInt] TypeBoolean),
    ("+.", TypeFunction [TypeFloat, TypeFloat] TypeFloat),
    ("-.", TypeFunction [TypeFloat, TypeFloat] TypeFloat),
    ("*.", TypeFunction [TypeFloat, TypeFloat] TypeFloat),
    ("/.", TypeFunction [TypeFloat, TypeFloat] TypeFloat),
    ("<.", TypeFunction [TypeFloat, TypeFloat] TypeBoolean)
  ]


libraryBuiltins :: [(Name, Type)]
libraryBuiltins = [
    ("exitSuccess", TypeFunction [] TypeUnit),
    ("exitFailure", TypeFunction [] TypeUnit),
    ("printInt", TypeFunction [TypeInt] TypeUnit),
    ("printChar", TypeFunction [TypeInt] TypeUnit),
    ("printFloat", TypeFunction [TypeFloat] TypeUnit),
    ("sin", TypeFunction [TypeFloat] TypeFloat),
    ("sqrt", TypeFunction [TypeFloat] TypeFloat)
  ]

File: rts/builtins.c
#include <stdlib.h>
#include <stdio.h>

double printInt(int i) {
    printf("%d", i);
    return 0.0;
}

double printDouble(double x) {
	printf("%f", x);
	return 0.0;
}

double printChar(double x) {
	printf("%c", (char)x);
	return 0.0;
}

double exitSuccess() {
	exit(0);
	return 0.0;
}

